nomad/deploymentwatcher/deployments_watcher.go

```diff
@@ -0,0 +1,414 @@
+package deploymentwatcher
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"sync"
+	"time"
+
+	"golang.org/x/time/rate"
+
+	"github.com/hashicorp/nomad/nomad/structs"
+)
+
+// DeploymentRaftEndpoints exposes the deployment watcher to a set of functions
+// to apply data transforms via Raft.
+type DeploymentRaftEndpoints interface {
+	// UpsertEvals is used to upsert a set of evaluations
+	UpsertEvals([]*structs.Evaluation) (uint64, error)
+
+	// UpsertJob is used to upsert a job
+	UpsertJob(job *structs.Job) (uint64, error)
+
+	// UpsertDeploymentStatusUpdate is used to upsert a deployment status update
+	// and potentially create an evaluation.
+	UpsertDeploymentStatusUpdate(u *structs.DeploymentStatusUpdateRequest) (uint64, error)
+
+	// UpsertDeploymentPromotion is used to promote canaries in a deployment
+	UpsertDeploymentPromotion(req *structs.ApplyDeploymentPromoteRequest) (uint64, error)
+
+	// UpsertDeploymentAllocHealth is used to set the health of allocations in a
+	// deployment
+	UpsertDeploymentAllocHealth(req *structs.ApplyDeploymentAllocHealthRequest) (uint64, error)
+}
+
+// DeploymentStateWatchers are the set of functions required to watch objects on
+// behalf of a deployment
+type DeploymentStateWatchers interface {
+	// Evaluations returns the set of evaluations for the given job
+	Evaluations(args *structs.JobSpecificRequest, reply *structs.JobEvaluationsResponse) error
+
+	// Allocations returns the set of allocations that are part of the
+	// deployment.
+	Allocations(args *structs.DeploymentSpecificRequest, reply *structs.AllocListResponse) error
+
+	// List is used to list all the deployments in the system
+	List(args *structs.DeploymentListRequest, reply *structs.DeploymentListResponse) error
+
+	// GetDeployment is used to lookup a particular deployment.
+	GetDeployment(args *structs.DeploymentSpecificRequest, reply *structs.SingleDeploymentResponse) error
+
+	// GetJobVersions is used to lookup the versions of a job. This is used when
+	// rolling back to find the latest stable job
+	GetJobVersions(args *structs.JobSpecificRequest, reply *structs.JobVersionsResponse) error
+
+	// GetJob is used to lookup a particular job.
+	GetJob(args *structs.JobSpecificRequest, reply *structs.SingleJobResponse) error
+}
+
+const (
```

Comment:
I would move the `const` and `var` blocks above the interfaces. I would rename the `EvalBatchDuration` as well since it's confusing with the `evalBatchDuration`.